script/search/2.0/search.inc.php

```diff
@@ -0,0 +1,357 @@
+  require_once(__DIR__ . '/../../../tools/db/db.php');
+
+  class KeyboardSearchResult {
+    const FILTER_SEARCH='search'; // default, search for the text string [any context]
+    const FILTER_ALL='all';       // return all possible results ?? TODO: check meaning of this
+    const FILTER_ID='id';         // filter by identifier per context, e.g. keyboard id, language tag, etc. [any context]
+    const FILTER_LEGACY='legacy'; // keyboards only, filter by a legacy id [CONTEXT_KEYBOARD]
+    const FILTER_LANGUAGE='language'; // search string is a language tag; [CONTEXT_KEYBOARD]
+    const FILTER_REGION='region';     // search string is a country identifier; [CONTEXT_LANGUAGE, later:CONTEXT_KEYBOARD]
+
+    const CONTEXT_KEYBOARD='keyboard';
+    const CONTEXT_LANGUAGE='language';
+    const CONTEXT_REGION='region';
+
+    public string $filter;   // none|id|legacy|language|region
+    public string $context;  // keyboard|language|region
+
+    public $text, $searchtext;
+
+    public int $pageNumber, $pageSize, $totalRows;
+    public $platform;
+
+    public $keyboards, $languages, $regions; // Only one of these can be populated at a time
+  }
+
+  class KeyboardSearch {
+    private $mssql;
+
+    const PAGESIZE = 100; //TODO: reduce to 10 once we support pagniation on keyman.com
+
+    function __construct($mssql) {
+      $this->mssql = $mssql;
+    }
+
+    function GetSearchMatches($context, $platform, $query, $pageNumber) {
+      $result = new KeyboardSearchResult();
+      $result->context = $context;
+      $result->pageSize = KeyboardSearch::PAGESIZE;
+      $result->pageNumber = $pageNumber;
+
+      if(in_array($platform, array('macos', 'windows', 'linux', 'android', 'ios'))) {
+        $result->platform = $platform;
+      } else if (in_array($platform, ['desktopWeb', 'mobileWeb'])) {
+        $result->platform = 'web';
+      }
+
+      $result->filter = KeyboardSearchResult::FILTER_SEARCH;
+      $result->regionmatch = false;
+
+      $textparts = explode(':', $query);
+
+      for($i = 0; $i < sizeof($textparts) - 1; $i++) {
+        $match = strtolower($textparts[$i]);
+        if(!strcmp($match, 'bcp47')) $result->filter = KeyboardSearchResult::FILTER_LANGUAGE;
+        else if(!strcmp($match, 'id')) $result->filter = KeyboardSearchResult::FILTER_ID;
+        else if(!strcmp($match, 'all')) $result->allmatch = true;
+        else if(!strcmp($match, 'region')) $result->regionmatch = true;
+        else if(!strcmp($match, 'legacy')) $result->filter = KeyboardSearchResult::FILTER_LEGACY;
+      }
+
+      $result->text = array_pop($textparts);
+      $result->searchtext = $result->text;
+
+      return $this->WriteSearchResults($result);
+    }
+
+    private function GetSearchQueries(KeyboardSearchResult $result) {
+      $result->searchtext = strip_tags($result->searchtext);
+
+      switch($result->context) {
+      case KeyboardSearchResult::CONTEXT_KEYBOARD:
+        return $this->LoadKeyboardSearch($result);
```

Comment:
So... we only get weighted results for this case?  (Noting that `LoadKeyboardSearch` is what implements the `match` object.)

Reply:
Yes, only keyboard searches are currently implemented. Probably going to strip out the other contexts per discussions earlier.